packages/beacon-node/src/chain/blocks/blockInput/blockInput.ts

```diff
@@ -0,0 +1,875 @@
+  protected bothPromise = createPromise<this>();
+
+  constructor(init: BlockInputInit) {
+    this.daRequirement = init.daRequirement;
+    this.timeBeginSec = init.timeBeginSec;
+    this.forkName = init.forkName;
+    this.slot = init.slot;
+    this.blockRootHex = init.blockRootHex;
+    this.parentRootHex = init.parentRootHex;
+  }
+
+  abstract addBlock(props: AddBlock<F>): void;
+
+  hasBlock(): boolean {
+    return this.state.blockStatus === BlockStatus.HasBlock;
+  }
+
+  getBlock(): SignedBeaconBlock<F> {
+    if (this.state.blockStatus !== BlockStatus.HasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISSING_BLOCK,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot getBlock from BlockInput without a block"
+      );
+    }
+    return this.state.block;
+  }
+
+  getBlockSource(): SourceMeta {
+    if (this.state.blockStatus !== BlockStatus.HasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISSING_BLOCK,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot getBlockSource from BlockInput without a block"
+      );
+    }
+    return this.state.source;
+  }
+
+  hasData(): boolean {
+    return this.state.daStatus === DAStatus.CompleteData;
+  }
+
+  hasBlockAndData(): boolean {
+    return this.state.blockStatus === BlockStatus.HasBlock && this.state.daStatus === DAStatus.CompleteData;
+  }
+
+  getLogMeta(): LogMetaBasic {
+    return {
+      blockRoot: prettyBytes(this.blockRootHex),
+      slot: this.slot,
+    };
+  }
+
+  getTimeComplete(): number {
+    if (this.state.blockStatus !== BlockStatus.HasBlock || this.state.daStatus !== DAStatus.CompleteData) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISSING_TIME_COMPLETE,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot getTimeComplete from BlockInput without a block and data"
+      );
+    }
+    return this.state.timeCompleteSec;
+  }
+
+  async waitForBlock(timeout: number, signal?: AbortSignal): Promise<SignedBeaconBlock<F>> {
+    if (this.state.blockStatus === BlockStatus.MissingBlock) {
+      return await withTimeout(() => this.blockPromise.promise, timeout, signal);
+    }
+    return this.state.block;
+  }
+  async waitForData(timeout: number, signal?: AbortSignal): Promise<TData> {
+    if (this.state.daStatus === DAStatus.IncompleteData) {
+      await withTimeout(() => this.dataPromise.promise, timeout, signal);
+    }
+    // each BlockInput implementation maintains its own repr of data
+    // so we just return the promise
+    return this.dataPromise.promise;
+  }
+  async waitForBlockAndData(timeout: number, signal?: AbortSignal): Promise<this> {
+    if (this.state.blockStatus === BlockStatus.MissingBlock || this.state.daStatus === DAStatus.IncompleteData) {
+      return await withTimeout(() => this.bothPromise.promise, timeout, signal);
+    }
+    return this;
+  }
```

Comment:
Can get rid of the `bothPromise` and just await a Promise.all and then return `this`

```suggestion
  async waitForBlockAndData(timeout: number, signal?: AbortSignal): Promise<this> {
    await withTimeout(() => Promise.all([this.blockPromise.promise, this.dataPromise.promise]), timeout, signal);
    return this;
  }
```